### ISSUE TITLE
Does OpenImageIO support nChannel images

### ISSUE THREAD

**Issue by StefanWoe** [OpenImageIO/oiio#1603]
Hi, just a question
can images (tiff, PSD and others) with more than 4 Channels be read into a memory buffer and written back using OpenImageIO? In the printing industry images with CMYK plus a number of spot color channels are quite common. Photoshop files must only be read, but TIFF must be written.

Is also BigTIFF supported?

Regards

Stefan

**Comment by lgritz** [comment]
Yes and yes.

**Comment by StefanWoe** [comment]
Thanks for your answer. In the meanwhile i found in the Documentation the following statement:

> Color space is grayscale or RGB. Non-spectral data, such as XYZ, CMYK, or YUV, are
converted to RGB upon reading.
(1.2 Simplifying Assumptions)

Doesnt that say that that even CMYK can not be handled?
Or is the Documentation incorrect?

**Comment by lgritz** [comment]
Ah, interesting point I hadn't considered. OIIO is fine with an arbitrary number of channels generally, but by default will try convert a CMYK representation to RGB.

Looking at the code, it seems that the PSD reader has a way to signal it to skip the CMYK->RGB conversion (by setting a special 'configuration' attribute called "psd:RawData" to be nonzero). But there does not appear to be a corresponding way to cause the TIFF reader to not do the transformation.

That seems like an oversight. Let me see if I can quickly propose a patch to address this. Stay tuned.

**Comment by StefanWoe** [comment]
Hi Igritz. That would be great help. Thanks!

**Comment by lgritz** [comment]
Proposed solution in #1605, when that gets approved, I'll merge it into master as well as the RB-1.7 release branch.

**Comment by StefanWoe** [comment]
Will this also work for multichannel (6, 8 or more channels?
Do you have a releaseplan (rough date) for RB_1.7?

**Comment by lgritz** [comment]
It has always worked fine for an arbitrary number of channels; the issue here was only that when the file was marked as PHOTOMETRIC_CMYK, it was auto-converting to RGB (and forcing 3 colors) without a way to say that you didn't want that conversion.

RB-1.7 is the *branch* that is the tip of 1.7 release development. You can always check out that branch tag and get the latest (well, once this patch is approved and I merge it). Generally, on the first of each month, I make an official release by adding a tag to the then-tip of RB-1.7, so that would be called "Release-1.7.11", for example.

So the short answer is that if you are cloning from GitHub and building yourself, you can have the fix as soon as I merge it (whether you are on 1.7 or using master), but Feb 1 is when an official release will happen that various compiled distributions are likely to pick up.

**Comment by StefanWoe** [comment]
As i see #1605 you already fixed the problem?

**Comment by lgritz** [comment]
I have proposed the patch of #1605 as the solution, and you are welcome to directly pull that patch from my repo (the lgritz GitHub user account) and try it on your end, but I have not yet merged it into the release codebase on the OpenImageIO account.

**Comment by lgritz** [comment]
Fix merged into master and RB-1.7. The next tagged release of 1.7 will be on 1 Feb.

**Comment by StefanAchm** [comment]
Have you tested the workflow with an nChannel-psd-file?
I tried to read an psd-file with 5 channels, but as a result i get 4 channels. (The command looks like this: "oiiotool --iconfig oiio:RawColor 1 --info -v multichannel.psd")

**Comment by lgritz** [comment]
Ah, it must be "multichannel mode". Looking at the code, it doesn't seem to support that currently, but I think that it should be easy to add support. Can you email me a (small) psd file with >4 channels that I can use as a test case?

**Comment by StefanAchm** [comment]
Also tried with multichannel-psd, but there i get 0 channels ...


I attached two files with each 5 channels (C,M,Y,K and 'channel5').

One (5channel.psd) is safed as CMYK, and the other (multichannel.psd) is safed as Multichannel.


________________________________
Von: Larry Gritz <notifications@github.com>
Gesendet: Donnerstag, 02. März 2017 21:42
An: OpenImageIO/oiio
Cc: StefanAchm; Comment
Betreff: Re: [OpenImageIO/oiio] Does OpenImageIO support nChannel images (#1603)


Ah, it must be "multichannel mode". Looking at the code, it doesn't seem to support that currently, but I think that it should be easy to add support. Can you email me a (small) psd file with >4 channels that I can use as a test case?

**Comment by StefanWoe** [comment]
Hello Igitz. Anything new about that?

**Comment by lgritz** [comment]
Working on it!

**Comment by lgritz** [comment]
#1641 is my proposed fix to allow reading of multichannel and cmyk (and grayscale, and fix 32 bits).
Please give it a try if you can, it would be nice to verify that it covers all your cases, @StefanWoe .

**Comment by StefanWoe** [comment]
I now made a test reading a tiff file with 10 channels:

```
ImageSpec config;
config.attribute("oiio:RawColor", 1);
mInputPtr = ImageInput::open(inFilePath, &config);
const ImageSpec &spec = mInputPtr->spec();
for(auto it : spec.channelnames)
        std::count << it << "\n";
```
Only 4 Channels are printed. 

Also spec.nchannels reports only 4 channels.

How can i retrieve all 10 channels?

**Comment by lgritz** [comment]
Can you put the file someplace where I can download it?

**Comment by StefanWoe** [comment]
You can download the file here:
https://www.file-upload.net/download-12408726/Test_Pantone_CS6_test.tif.html

**Comment by lgritz** [comment]
Which OIIO release are you using?

Instead of checking the channel names, what happens if you check spec.nchannels?

For me, it says it's a 10 channel image, but it badly botches the channel names. That's definitely a bug. It seems that it's assuming that if the color model is "separated", it represents a CMYK image and sets channel names accordingly (that is, wrongly). But it does appear to report the right number of channels (in spec.nchannels) and have all the pixel data.

A apologize for all the rough edges. In the VFX world, arbitrary number of channels is very common, but always for "spectral" or "data" images (RGB, plus lots of "data" channels like encoding of surface normals or other quantities about the geometry or shading). We in VFX rarely if ever see CMYK style images, so those parts of the code are a little under-tested and we're still squashing some of the edge cases (like in this case, it worked with straight-up 4-ink CMYK, but never was never tested with n-inks).

I will prepare a fix. Thanks for the test case!

**Comment by StefanWoe** [comment]
We are using currently:
#define OIIO_VERSION_MAJOR 1
#define OIIO_VERSION_MINOR 8
#define OIIO_VERSION_PATCH 3

This is a very important project for us.  Appreciate you support!

**Comment by lgritz** [comment]
Proposed fix is PR #1658 

I don't have a good way to test or display a color-separated 10-ink image. :-) Can you give a try on your end and make sure it seems to behave properly?

**Comment by StefanWoe** [comment]
1) I locally merged this PR and everything looks pretty now. Thanks a lot! When will this be merged into the master?
2) I also found that in the File i sent you the additional channels were empty - for reasons i can not reproduce. I uploaded a corrected file at https://www.file-upload.net/download-12419644/Test_Pantone_CS6_test.tif.html
3) Im still struggeling with psd files. Will return with that.

**Comment by StefanWoe** [comment]
I still found a remaining problem with multichannel tiff:
When i read the last file is sent you spec.channelnames contains the correct colors:
[Cyan]
[Magenta]
[Yellow]
[Black]
[PANTONE 877 C]
[PANTONE Cool Gray 10 C]
[PANTONE Warm Red C]
[PANTONE 280 C]
[PANTONE 7466 C]
[PANTONE 360 C]

When i write this file again as a 10 channel tiff file, the content is inverted (but still is correct), Channel Names are:
[R]
[G]
[B]
[A]
there is no Name for channels 5-10
I uploaded this file to
https://www.file-upload.net/download-12419669/Test_Pantone_CS6_test_rw.tif.html

**Comment by lgritz** [comment]
Yes, I found a couple of minor bugs that caused the channel names to get mangled on the way out.

Can you try again with the updated patch of #1658?

I'm not sure what you mean by "content is inverted." Can you clarify?

**Comment by lgritz** [comment]
Does the new fix in #1658 work now for you?

**Comment by StefanWoe** [comment]
Will need some days to reply.

**Comment by StefanWoe** [comment]
I merged in the changes - but i can not find a difference concerning the channel names.

**Comment by lgritz** [comment]
Here's what I get:

before patch #1658
-------------------
**$ oiiotool -v -info Test_Pantone_CS6_test.tif**
(let default transformations happen, it reports RGB which is totally wrong)
```
Reading Test_Pantone_CS6_test.tif
Test_Pantone_CS6_test.tif :  642 x  888, 3 channel, uint8 tiff
    channel list: R, G, B
    oiio:BitsPerSample: 8
    Orientation: 1 (normal)
    XResolution: 72
    YResolution: 72
    tiff:PhotometricInterpretation: 5
    tiff:ColorSpace: "CMYK"
    tiff:PlanarConfiguration: 1
    planarconfig: "contig"
    tiff:Compression: 5
    compression: "lzw"
    tiff:RowsPerStrip: 1
    PixelAspectRatio: 1
```
**$ oiiotool -v -info -iconfig oiio:rawcolor 1 Test_Pantone_CS6_test.tif**
(ask for raw color, it reports 10 channels but botches the channel names, thinking it's a "CMYK" image plus extra data channels)
```
Reading Test_Pantone_CS6_test.tif
Test_Pantone_CS6_test.tif :  642 x  888, 10 channel, uint8 tiff
    channel list: C, M, Y, K, unknown, unknown, unknown, unknown, unknown, unknown
    oiio:BitsPerSample: 8
    Orientation: 1 (normal)
    XResolution: 72
    YResolution: 72
    tiff:PhotometricInterpretation: 5
    tiff:ColorSpace: "CMYK"
    oiio:ColorSpace: "CMYK"
    tiff:PlanarConfiguration: 1
    planarconfig: "contig"
    tiff:Compression: 5
    compression: "lzw"
    tiff:RowsPerStrip: 1
    PixelAspectRatio: 1
```

after patch #1658
-----------------
**$ oiiotool -v -info Test_Pantone_CS6_test.tif**
(default behavior is now to report back all the channels, including their proper ink names, and knows it's a "color separated" image rather than thinking it's strictly CMYK)
```
Reading Test_Pantone_CS6_test.tif
Test_Pantone_CS6_test.tif :  642 x  888, 10 channel, uint8 tiff
    channel list: Cyan, Magenta, Yellow, Black, PANTONE 877 C, PANTONE Cool Gray 10 C, PANTONE Warm Red C, PANTONE 280 C, PANTONE 7466 C, PANTONE 360 C
    oiio:BitsPerSample: 8
    Orientation: 1 (normal)
    XResolution: 72
    YResolution: 72
    tiff:PhotometricInterpretation: 5
    tiff:ColorSpace: "color separated"
    oiio:ColorSpace: "color separated"
    tiff:PlanarConfiguration: 1
    planarconfig: "contig"
    tiff:Compression: 5
    compression: "lzw"
    tiff:RowsPerStrip: 1
    PixelAspectRatio: 1
```

**Comment by StefanWoe** [comment]
Excuse me, i was not precise enough. The colors are read correctly from a 10 Channel TIFF, but they are not saved when writing the same buffer to a new file.

**Comment by lgritz** [comment]
**$ oiiotool Test_Pantone_CS6_test.tif -o out.tif**
**$ oiiotool -info -v out.tif**
```
Reading out.tif
out.tif              :  642 x  888, 10 channel, uint8 tiff
    channel list: Cyan, Magenta, Yellow, Black, PANTONE 877 C, PANTONE Cool Gray 10 C, PANTONE Warm Red C, PANTONE 280 C, PANTONE 7466 C, PANTONE 360 C
    oiio:BitsPerSample: 8
    Orientation: 1 (normal)
    XResolution: 72
    YResolution: 72
    Software: "OpenImageIO 1.8.4dev : oiiotool Test_Pantone_CS6_test.tif -o out.tif"
    DateTime: "2017:04:19 13:34:05"
    tiff:PhotometricInterpretation: 5
    tiff:ColorSpace: "color separated"
    oiio:ColorSpace: "color separated"
    tiff:PlanarConfiguration: 1
    planarconfig: "contig"
    tiff:Compression: 5
    compression: "lzw"
    tiff:RowsPerStrip: 1
    PixelAspectRatio: 1
```

Looks ok to me.

How can I reproduce the problem?

**Comment by StefanWoe** [comment]
I just commented out the following line:
src\tiff.imageio\tiffoutput.cpp 449:
    // if (m_photometric == PHOTOMETRIC_RGB) 

Then the stored channel names seem to be correct. Without that inknames in the tiff file are not set. The only single call to 
TIFFSetField (m_tif, TIFFTAG_INKNAMES, 
is inside that case.

Before saving i set:
            spec.attribute("tiff:ColorSpace", "color separated");
wich seems correct to me?

The problem has its root maybe in Line 396:
    m_photometric = (m_spec.nchannels > 1 ? PHOTOMETRIC_RGB : PHOTOMETRIC_MINISBLACK);

**Comment by lgritz** [comment]
Can we move this discussion to #1658, so it's associated with the code?
